### PR TITLE
SAK-52724 Tests & Quizzes Modernize PDF export for assessments

### PR DESCRIPTION
**Pull request by st-manu**
https://sakaiproject.atlassian.net/browse/SAK-52724

## Summary by CodeRabbit

* **New Features**
  * Streamlined assessment printing with configurable settings, PDF.js preview, and downloadable PDFs.
  * Added support for student report PDFs, answer keys, feedback, attachments, media, and mathematical notation.
* **Bug Fixes**
  * Improved PDF export reliability and handling of missing assessment data.
  * Enhanced rendering consistency across question types and localized content.
* **Tests**
  * Added end-to-end coverage for print preview and answer-key behavior.
  * Added unit tests for PDF mapping, content rendering, and LaTeX parsing.

### CODE REVIEW

File: samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/evaluation/ExportAction.java [hunk 1]
```diff
@@ -85,15 +87,49 @@ public class ExportAction implements ActionListener {
 
     private ContentHostingService contentHostingService = org.sakaiproject.content.cover.ContentHostingService.getInstance();
 	private static final ResourceLoader rb = new ResourceLoader("org.sakaiproject.tool.assessment.bundle.AuthorMessages");
-	private Color gray = new Color(221, 219, 219);
-	private Color grayLight = new Color(231, 228, 228);
-	private Color green = new Color(9, 215, 71);
-	private Font boldFont = new Font(Font.TIMES_ROMAN, 13, Font.BOLD);
+	private static final ResourceLoader rbEval = new ResourceLoader("org.sakaiproject.tool.assessment.bundle.EvaluationMessages");
```
[the remark this answers is not in the document]
Comment (st-manu) [reply samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/evaluation/ExportAction.java]: Incorrect. The translation strings already exist:

src\java\org\sakaiproject\tool\assessment\bundle\AuthorMessages.properties
`no_answer.text = Unanswered`

File: samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/evaluation/ExportAction.java [hunk 2]
```diff
@@ -104,154 +143,311 @@ public void processAction(ActionEvent ae) {
...
+			DecimalFormat twoDecimalsFormat = new DecimalFormat("0.00");
+
+			// Student name
+			Paragraph studentNameParagraph = new Paragraph(studentScoreBean.getStudentName(), getFontWithColor(TITLE_FONT, TEXT_PRIMARY));
+			studentNameParagraph.setSpacingBefore(5f);
+			studentNameParagraph.setSpacingAfter(5f);
 			document.add(studentNameParagraph);
+
+			// Student email
 			if (studentScoreBean.getEmail() != null && !StringUtils.equals(studentScoreBean.getEmail(), "")) {
```
Comment (ottenhoff) [inline samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/evaluation/ExportAction.java]: StringUtils.isNotBlank

**Comment by ottenhoff** [comment]
I support this, and I'm happy to contribute to it. Here are some comments from my agent on how to improve this:


  - Blocker: samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/evaluation/ExportAction.java:140 is now both JSF action and full PDF renderer. processAction owns response handling, site lookup, summary rendering, all question-type rendering, comments, feedback, and styling from roughly lines 140-622. The file grows from 1,013 to 1,284 lines. Can we decompose this first? The listener should gather beans/headers and delegate to a focused submission PDF renderer/model.

  - Blocker: samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/evaluation/ExportAction.java:322 adds more spaghetti question-type flow. Each item falls through fill-in, essay, upload/audio, matrix, imagemap, answer-
    loop, matching-loop, correct-response exclusions, and feedback branches. The long negative condition at samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/evaluation/ExportAction.java:587 is the smell. Use a single `renderQuestion(...)` dispatch with per-type renderers/helpers so each type owns its body, answers, correct response, and feedback behavior.

  - High: the rewrite appears to drop overall student comments. The old code rendered `studentScoreBean.getComments()` before question details; the new flow jumps into per-part rendering at samigo/samigo-app/src/java/org/sakaiproject/tool/
    assessment/ui/listener/evaluation/ExportAction.java:225 and only emits per-item comments/feedback at samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/evaluation/ExportAction.java:591. If that omission is not intentional, it is a behavior regression hidden inside a layout change.

  - High: samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/evaluation/ExportAction.java:841 is being patched with lastValidIndex instead of being simplified. The parser still emits `text.substring(0, latexInitIndex)` inside each loop, which makes multiple math fragments hard to reason about and likely to duplicate prefix text. Extract a pure cursor-based parseLatexChunks(text) and test no math, one math fragment, multiple fragments, and unclosed delimiters.

  - Medium: no tests were added for either the null MatchingBean.getIsCorrect() fix at samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/evaluation/ExportAction.java:559 or the large PDF export rewrite. After extracting the renderer/parser seam, add focused public-path tests for null matching correctness, overall comments, and representative question-type rendering.

  - Medium: the new site lookup at samigo/samigo-app/src/java/org/sakaiproject/tool/assessment/ui/listener/evaluation/ExportAction.java:188 reaches into SiteService from the render path and catches broad Exception. Build site title once as optional export context, using the existing Samigo site-helper pattern, and avoid swallowing unexpected failures with only `e.getMessage()`.

**Comment by st-manu** [comment]
@ottenhoff

do you think it would make sense to unify the assessment printing flow (which is currently rendered as html and then converted to pdf) with the pdf generation for the student's assessment with responses?

it seems more consistent to me if both follow the same flow and design, since both features ultimately generate a pdf of the assessment, with the only difference being that one includes the student's responses

if you have a chance, could you take a look at the implementation? if you think it makes sense, feel free to make any changes you think would improve it

**Comment by ottenhoff** [comment]
> do you think it would make sense to unify the assessment printing flow (which is currently rendered as html and then converted to pdf) with the pdf generation for the student's assessment with responses?

Yes, great idea.

> 
> it seems more consistent to me if both follow the same flow and design, since both features ultimately generate a pdf of the assessment, with the only difference being that one includes the student's responses

I agree

> 
> if you have a chance, could you take a look at the implementation? if you think it makes sense, feel free to make any changes you think would improve it

I will look today

**Comment by ottenhoff** [comment]
### Blocker: Printable matrix questions lose their grid

- **Where:** `AssessmentPdfQuestionRenderer.java:77-83,199-202`
- **Issue:** `MATRIX_CHOICES_SURVEY` is treated as an ordinary choice question. The snapshot contains row labels, column labels, and matrix rows, but the printable renderer never consumes them.
- **Problematic code:**

```java
private static final Set<Long> PRINT_CHOICE_TYPES = Set.of(
        // ...
        TypeIfc.MATRIX_CHOICES_SURVEY);

if (PRINT_CHOICE_TYPES.contains(questionType)) {
    renderPrintChoiceAnswers(document, context, question, questionType, printSettings);
    return;
}
```

- **Proof:** I generated a printable model containing a matrix with row `Row A` and columns `Low`/`High`. Extracted PDF text contained only:

```text
Question 1 / 1
Rate each row
```

The row and both columns were absent. The previous JSP rendered a proper row/column matrix.
- **Why it matters:** Printed matrix assessments are unusable because students cannot see the response grid.
- **Fix direction:** Route this type to a dedicated printable matrix renderer using `getMatrixRows()`, `getColumnLabels()`, and `getColumnIndexes()`. Also preserve the matrix comment prompt currently available through `ItemContentsBean`.

```java
if (Objects.equals(questionType, TypeIfc.MATRIX_CHOICES_SURVEY)) {
    renderPrintMatrix(document, context, question);
    return;
}
```

### Should Fix: Rich-text feedback is emitted as literal HTML

- **Where:** `AssessmentPdfQuestionRenderer.java:252-254,370-388`
- **Issue:** Feedback fields are formatted HTML, but the renderer inserts them directly into `Chunk` or `Paragraph`.
- **Problematic code:**

```java
keyParagraph.add(new Chunk(
        question.getCorrectItemFeedback(),
        fontWithColor(context.smallFont(), TEXT_PRIMARY)));
```

- **Proof:** Rendering `<p>Good <strong>job</strong></p>` produces this visible PDF text:

```text
Correct Feedback: <p>Good <strong>job</strong></p>
```

The previous pipeline converted and parsed formatted feedback.
- **Why it matters:** Formatting, entities, images, and mathematics in answer-level and item-level feedback are broken.
- **Fix direction:** Render feedback through the existing HTML-aware content helper, ideally into a styled cell beneath the label. At minimum, convert it to clean text instead of exposing markup.

```java
PdfPCell feedbackCell = contentHelper.createInfoBoxCell(...);
contentHelper.populateCellWithHtml(
        feedbackCell,
        question.getCorrectItemFeedback(),
        context.smallFont(),
        context.isMathJaxEnabled(),
        context.getFontSizeSetting());
```

**Review by ottenhoff — APPROVED** [review]
